src/libs/safe-path/src/safe_join.rs

```diff
@@ -0,0 +1,303 @@
+
+            // Perform the checks
+            assert_eq!(&result, Path::new(t.result), "{}", msg);
+        }
+    }
+
+    #[test]
+    fn test_scoped_resolve() {
+        // create temporary directory to emulate container rootfs with symlink
+        let rootfs_dir = tempdir().expect("failed to create tmpdir");
+        let rootfs_path = rootfs_dir.path();
+
+        let tests = [
+            TestData {
+                name: "normal path",
+                rootfs: rootfs_path,
+                unsafe_path: "a/b/c",
+                result: "a/b/c",
+            },
+            TestData {
+                name: "path with .. at beginning",
+                rootfs: rootfs_path,
+                unsafe_path: "../../../a/b/c",
+                result: "a/b/c",
+            },
+            TestData {
+                name: "path with complex .. pattern",
+                rootfs: rootfs_path,
+                unsafe_path: "../../../a/../../b/../../c",
+                result: "c",
+            },
+            TestData {
+                name: "path with .. in middle",
+                rootfs: rootfs_path,
+                unsafe_path: "/usr/bin/../../bin/ls",
+                result: "bin/ls",
+            },
+            TestData {
+                name: "path with . and ..",
+                rootfs: rootfs_path,
+                unsafe_path: "/usr/./bin/../../bin/./ls",
+                result: "bin/ls",
+            },
+            TestData {
+                name: "path with . at end",
+                rootfs: rootfs_path,
+                unsafe_path: "/usr/./bin/../../bin/./ls/.",
+                result: "bin/ls",
+            },
+            TestData {
+                name: "path try to escape by ..",
+                rootfs: rootfs_path,
+                unsafe_path: "/usr/./bin/../../../../bin/./ls/../ls",
+                result: "bin/ls",
+            },
+            TestData {
+                name: "path with .. at the end",
+                rootfs: rootfs_path,
+                unsafe_path: "/usr/./bin/../../bin/./ls/..",
+                result: "bin",
+            },
+            TestData {
+                name: "path ..",
+                rootfs: rootfs_path,
+                unsafe_path: "..",
+                result: "",
+            },
+            TestData {
+                name: "path .",
+                rootfs: rootfs_path,
+                unsafe_path: ".",
+                result: "",
+            },
+            TestData {
+                name: "path /",
+                rootfs: rootfs_path,
+                unsafe_path: "/",
+                result: "",
+            },
+            TestData {
+                name: "empty path",
+                rootfs: rootfs_path,
+                unsafe_path: "",
+                result: "",
+            },
+        ];
+
+        exec_tests(&tests);
+    }
+
+    #[test]
+    fn test_scoped_resolve_invalid() {
+        scoped_resolve(".a", ".").unwrap_err();
```

Comment:
Yes, it's testing the case where `root` is not an absolute path. Originally I have a check for that case, but found it cannot be reached when trying to improve code coverage. It turns out that `fs::canicanize()` already done that work for us, because it returns "the canonical, absolute form of a path with all intermediate components normalized and symbolic links resolved."

Reply:
That doesn't seem like the same thing to me.  `canonicalize()` *converts* a relative path into an absolute path; that doesn't mean it won't work when given a relative path, so this test doesn't seem right.